### PR TITLE
Refactor bayesianapproximator

### PR DESCRIPTION
**Pull request by dchui1**

### CODE REVIEW

File: BNN.py [hunk 1]
```diff
@@ -47,15 +53,20 @@ def new_bnn(log_divergence_weight=-10, prior_stddev=1, residual_weight=0.1):
...
 
 # run model
 f = new_bnn()
 f.compile(optimizer=tf.keras.optimizers.Adam(1e-2), loss=tf.losses.huber_loss)
 history = f.fit(x, y, batch_size=32, epochs=1000, verbose=1)#, class_weight=W)
+# print("history", history)
+#
+# print("Finished compiling")
+# print(type(f))
+# print(type(x))
```
Comment (ParashRahman) [inline BNN.py:69]: I recommend deleting this
```suggestion
# print("history", history)
#
# print("Finished compiling")
# print(type(f))
# print(type(x))
```

File: BNN.py [hunk 2]
```diff
@@ -67,12 +78,13 @@ def new_bnn(log_divergence_weight=-10, prior_stddev=1, residual_weight=0.1):
   predictions_list_trainset.append(y)
 
 predictions_list_trainset = np.asarray(predictions_list_trainset)
-print(predictions_list_trainset.shape)
-print(predictions_list_trainset.mean())
-print((predictions_list_trainset.std(0, ddof=1)).mean())
+
+print("Predictions shape", predictions_list_trainset.shape)
+print("Predictions mean", predictions_list_trainset.mean())
+print("Predictions shape", predictions_list_trainset.std(0, ddof=1))
```
Comment (ParashRahman) [inline BNN.py:84]: I think you meant "Predictions std"

```suggestion
print("Predictions shape", predictions_list_trainset.std(0, ddof=1))
```

File: q_learning.py [hunk 3]
```diff
@@ -33,6 +36,15 @@ def update(self, s, sp, r, a, done):
...
 
+
+# class Agent:
+#     def __init__(self):
+#         self.B = None
+#
+#     def start(self, obs):
+#         self.next_action = self.policy(obs)
+#         return self.next_action
+
```
Comment (ParashRahman) [inline q_learning.py:47]: Recommend deleting these comments

File: BNNApproximation.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        # input = np.array(state_dimensions + self.convert_action(self.num_actions))
+        # self.history = self.bnn.fit(input, 0, batch_size=32,
+        #     epochs=1000, verbose=1)
+
+
+    def update_stats(self, s, a, val=0.0):
+        if self.dimensions.shape != s.shape or not a <= self.num_actions:
+
+            raise ValueError("Invalid value to update stats", s, a, val)
+        input = np.concatenate((s, self.convert_action(a)))
```
Comment (ParashRahman) [inline BNNApproximation.py:24]: Change input to not a Python keyword

File: BNNApproximation.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        #TODO refactor this back out into the BNN method
+        self.bnn.compile(optimizer=tf.keras.optimizers.Adam(1e-2),
+            loss=tf.losses.huber_loss)
+        # input = np.array(state_dimensions + self.convert_action(self.num_actions))
+        # self.history = self.bnn.fit(input, 0, batch_size=32,
+        #     epochs=1000, verbose=1)
+
+
+    def update_stats(self, s, a, val=0.0):
+        if self.dimensions.shape != s.shape or not a <= self.num_actions:
```
Comment (ParashRahman) [inline BNNApproximation.py:21]: Should be checking 'a < 1' as well. Recommend a > num_actions for readibility

File: BNNApproximation.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        print(input.shape)
+
+        self.bnn.fit(np.array([input, input]), np.array([val, val]), batch_size=32, epochs=100, verbose=1)
+
+
+    def sample(self, s, a, n):
+        if self.dimensions.shape != s.shape or not a <= self.num_actions:
+            raise ValueError("Invalid value to sample", s, a)
+
+        input = np.concatenate((s, self.convert_action(a)))
```
Comment (ParashRahman) [inline BNNApproximation.py:34]: Change from Python keyword input

File: BNNApproximation.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+from BNN import new_bnn
+import tensorflow as tf
+from bayesianapproximator import BayesianApproximator
```
Comment (ParashRahman) [inline BNNApproximation.py:4]: Why are they called bayesianApproximators and BayesianApproximator?

File: q_learning.py [hunk 8]
```diff
@@ -197,23 +184,38 @@ def plotRewards(ax, rewards, stderr, label):
...
 
-plt.legend()
-plt.show()
+# plt.legend()
+# plt.show()
+
+
+
+def parse_args():
+    parser = argparse.ArgumentParser("Reinforcement Learning experiments for multiagent environments")
```
Comment (andnp) [inline q_learning.py:212]: I'm getting errors saying that `argparse` is not defined. Does this need to be imported?

File: q_learning.py [hunk 9]
```diff
@@ -197,23 +184,38 @@ def plotRewards(ax, rewards, stderr, label):
...
 # plotRewards(ax, rewards, stderr, 'Optimal')
+#
+# (rewards, stderr) = averageOverRuns(Q, env, 1)
+# plotRewards(ax, rewards, stderr, 'Q epsilon=0.1')
 
 # (rewards, stderr) = averageOverRuns(Q, env, 1)
 # plotRewards(ax, rewards, stderr, 'Q epsilon=0.1')
 
 (rewards, stderr) = averageOverRuns(QRewardValueFunction, env, 5)
 plotRewards(ax, rewards, stderr, 'QReward value-function')
```
Comment (andnp) [inline q_learning.py:204]: `ax` isn't defined here

File: q_learning.py [hunk 10]
```diff
@@ -171,17 +161,14 @@ def averageOverRuns(Agent, env, runs = 20):
...
     (steps, r) = runExperiment(env, 1000, agent)
     rewards.append(r)
     total_steps.append(steps)
 
-  # mean_value = np.mean(agent.Q, 1).reshape(30, 30)
-  # plt.imshow(mean_value, cmap='hot', interpolation='nearest')
-  # plt.gca().invert_yaxis()
-  # plt.show()
+
+  print("Values of table afterwards: ",  bayesianApproximator.B)
```
Comment (andnp) [inline q_learning.py:171]: Looks like the `bayesianApproximator` object does not have a `B` attribute.

**Pull request merged**